### PR TITLE
Serverless services

### PR DESCRIPTION
**Pull request by javfg**
This PR adds Serverless services to Reva. It is the first step before the big PR for the Notifications Service.

This adds a new type of service which doesn't have a listener (neither http or grp). As the Notifications service is communicating via NATS only with other backend services, it doesn't need to be exposed.

It could later on be used to implement other services which don't need http/grp endpoints, like metrics.

### CODE REVIEW

**Comment by labkode** [comment]
@javfg can you add an example of how the config would look like for an example service?

**Comment by gmgigi96** [comment]
This is awesome, we could use the "serverless" services for running some internal reva procedures!

Regarding the code, I have a comment regarding the management of the services. In the HTTP and GRPC services, revad is handling properly the graceful shutdown when a signal is received (check https://github.com/cs3org/reva/blob/master/cmd/revad/internal/grace/grace.go#L258), so every service does not have to implement its own signal handler.

Also, how do you deal with errors in a single service? We don't have any way of propagating this to the Serverless struct. If we have a way then, do we need to restart the single service? Do we need to panic the entire reva?

**Comment by javfg** [comment]
> This is awesome, we could use the "serverless" services for running some internal reva procedures!
> 
> Regarding the code, I have a comment regarding the management of the services. In the HTTP and GRPC services, revad is handling properly the graceful shutdown when a signal is received (check https://github.com/cs3org/reva/blob/master/cmd/revad/internal/grace/grace.go#L258), so every service does not have to implement its own signal handler.

I looked into this initially and thought about adding some way to deal with the shutdown for serverless in `grace.go` too, but in the end I did not because I'm not sure if there is a way to deal with them in a unified way.

Maybe we could discuss this during next week? I thought perhaps using context (which serverless services don't have right now), and maybe using `context.WithCancel` to start the services?

> Also, how do you deal with errors in a single service? We don't have any way of propagating this to the Serverless struct. If we have a way then, do we need to restart the single service? Do we need to panic the entire reva?

Not dealing with them at all (a broken service crashes the whole Reva instance), and indeed this is something that is missing too. I'm not sure about how we handle them in rhttp/rgrpc. Another thing to discuss, I could use some input.

**Comment by javfg** [comment]
> @javfg can you add an example of how the config would look like for an example service?

Sure! Here is the config for the notifications service:

```toml
[serverless.services.notifications]
nats_address = ""
nats_token = "TOKEN"
nats_template_subject = "reva-notifications-template"
nats_notification_subject = "reva-notifications-notification"
nats_trigger_subject = "reva-notifications-trigger"
storage_driver = "sql"
grouping_interval = 5
grouping_maxsize = 5
```

**Comment by labkode** [comment]
@javfg thanks!

@gmgigi96 has a good point. Even if they are server less, I expect that these services will create network connections to upstream services and maybe open local files. These cases should be handled gracefully if possible. 

Adding the following methods shouldn't be difficult. We can split he interface into two: Gracer and Listener
```
Stop() error
GracefulStop() error
```

File: pkg/rserverless/rserverless.go [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+// Start starts the serverless service collection.
+func (s *Serverless) Start() error {
+	return s.registerServices()
+}
+
+func (s *Serverless) registerServices() error {
+	for svcName := range s.conf.Services {
+		if s.isServiceEnabled(svcName) {
+			newFunc := Services[svcName]
+			svc, err := newFunc(s.conf.Services[svcName], &s.log)
```
Comment (gmgigi96) [inline pkg/rserverless/rserverless.go:85]: Here I would pass a logger containing the service name

File: pkg/rserverless/rserverless.go [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+			if err != nil {
+				return errors.Wrapf(err, "serverless service %s could not be initialized", svcName)
+			}
+
+			go func() {
+				svc.Start()
```
Comment (gmgigi96) [inline pkg/rserverless/rserverless.go:91]: this can be simplified with just `go svc.Start()`

File: pkg/rserverless/rserverless.go [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+	_, ok := Services[svcName]
+	return ok
+}
+
+// Start starts the serverless service collection.
+func (s *Serverless) Start() error {
+	return s.registerServices()
+}
+
+func (s *Serverless) registerServices() error {
```
Comment (gmgigi96) [inline pkg/rserverless/rserverless.go:81]: call this `registerAndStartServices`

File: pkg/rserverless/rserverless.go [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+func New(m interface{}, l zerolog.Logger) (*Serverless, error) {
+	conf := &config{}
+	if err := mapstructure.Decode(m, conf); err != nil {
+		return nil, err
+	}
+
+	n := &Serverless{
+		conf:     conf,
+		log:      l,
+		services: map[string]Service{},
```
Comment (gmgigi96) [inline pkg/rserverless/rserverless.go:66]: Why do you need this map?

File: cmd/revad/runtime/runtime.go [hunk 5]
```diff
@@ -141,13 +149,21 @@ func initServers(mainConf map[string]interface{}, log *zerolog.Logger) map[strin
...
-		os.Exit(1)
-	}
 	return servers
 }
 
+func initServerless(mainConf map[string]interface{}, log *zerolog.Logger) *rserverless.Serverless {
+	if isEnabledServerless(mainConf) {
+		serverless, err := getServerless(mainConf["serverless"], log)
+		if err != nil {
+			log.Error().Err(err).Msg("error")
```
Comment (gmgigi96) [inline cmd/revad/runtime/runtime.go:159]: you should abort here

**Review by gmgigi96 — CHANGES_REQUESTED** [review]

File: pkg/rserverless/rserverless.go (hunk 1 above)
Comment (javfg) [reply pkg/rserverless/rserverless.go:85]: Good idea! I'll also add this to `rhttp`. gRPC services each handle their own logger, it seems.

File: pkg/rserverless/rserverless.go (hunk 4 above)
Comment (javfg) [reply pkg/rserverless/rserverless.go:66]: Indeed, this requires clarification.

This is there to follow the other types of services ([`rhttp`](https://github.com/cs3org/reva/blob/master/pkg/rhttp/rhttp.go#L69) and [`rgrpc`](https://github.com/cs3org/reva/blob/master/pkg/rgrpc/rgrpc.go#L122C5)). It is unused right now, but I'll make use of it for the graceful stop.

File: pkg/rserverless/rserverless.go [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+				}
+			} else {
+				s.log.Info().Msgf("service %s stopped", svcName)
+				stoppedServices[svcName] = true
+			}
+		}(svcName, svc)
+	}
+
+	count := 9 // one second less than the grace watcher deadlne
+	ticker := time.NewTicker(time.Second)
```
Comment (gmgigi96) [inline pkg/rserverless/rserverless.go:120]: I don't understand this for-loop, you can do the same using a `context.WithDeadline`, passed also to the `Close` method (see the previous comment)

File: pkg/rserverless/rserverless.go [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+			}
+
+			go svc.Start()
+
+			s.services[svcName] = svc
+
+			s.log.Info().Msgf("serverless service enabled: %s", svcName)
+		} else {
+			message := fmt.Sprintf("serverless service %s does not exist", svcName)
+			return errors.New(message)
```
Comment (gmgigi96) [inline pkg/rserverless/rserverless.go:169]: Use `fmt.Errorf`

**Review by gmgigi96 — CHANGES_REQUESTED** [review]

**Review by gmgigi96 — APPROVED** [review]

**Pull request merged**